pkg/parser/convert.go

```diff
@@ -19,136 +25,47 @@ type Paragraph struct {
 	Sentences []Sentence
 }
 
-type Document struct {
-	Tags       []string
-	Title      []string
-	Paragraphs []Paragraph
+type Attributes struct {
+	Columns []string
+	Values  map[string][]string
 }
 
-// Convert takes the bytes from a file, and converts those bytes to an intermediate format by breaking the document down into
-// paragraphs, sentences, and  words. For now punctuation and words are roughly equivalent. This is somewhat problematic in that stop != stop.
-// however it makes the implementation significantly easier for now. Saving punctuation forces the representationt o also
-// keep tracing spaces for example "some - text" versus "some-text" versus "12/30/2020".
-func Convert(unprocessed []byte) Document {
-
-	d := Document{
-		Paragraphs: make([]Paragraph, 0, 0)}
-
-	cursor := 0
-	length := len(unprocessed)
-	for cursor < length {
-		switch unprocessed[cursor] {
-		case ' ', '\t', '\n', '\f', '\r':
-			cursor++
-		default:
-			paragraph, cursorAfterParagraph := readParagraph(unprocessed, cursor)
-
-			d.Paragraphs = append(d.Paragraphs, paragraph)
-
-			cursor = cursorAfterParagraph
-		}
-	}
-
-	return d
+type Document struct {
+	Attributes Attributes
+	Title      string
+	Paragraphs []Paragraph
 }
 
-func readParagraph(text []byte, position int) (Paragraph, int) {
-	p := Paragraph{
-		Sentences: []Sentence{},
-	}
-	for position < len(text) {
-		switch text[position] {
-
-		// we assume we hit another paragraph even without punctuation
-		case '\n', '\f', '\r':
-			return p, position
-
-		// skip spaces
-		case '\t', ' ':
-			position += 1
-
-		// otherwise read the sentence
-		default:
-			sentence, newPosition := readSentence(text, position)
-
-			p.Sentences = append(p.Sentences, sentence)
-
-			position = newPosition
-		}
-	}
-
-	return p, position
+type Parser interface {
+	Convert([]byte) (Document, error)
 }
 
-func readSentence(text []byte, position int) (Sentence, int) {
-	s := Sentence{
-		Parts: []Word{},
+// GetParser returns the correct parser for the filetype or error if there is no correct parser
+// TODO may be worth adding an extra method to the Parser interface that returns a list of parseable documents
+// then parsers can automatically declare themselves. On the other hand we would then need to keep around
+// instances of the different parsers. The current approach uses a bit less state, with a lot of short lived
+// parsers. This feels appropriate since parsers should never have state that extends past the lifetime of
+// the document that they have just finished parsing.
```

Comment:
You know honestly that TODO is a bit outdated anyways. But that's good to know. I more do it for the visual aspect of it, since it looks weird when just one line is highlighted for me.